### PR TITLE
Add frontrun protection using witness and secret

### PR DESCRIPTION
**Pull request by Agusx1211**

### CODE REVIEW

File: contracts/UniswapEX.sol [hunk 1]
```diff
@@ -61,7 +62,27 @@ contract UniswapEX {
...
+            uint256 fee,
+            address payable owner,
+            ,
+            address witness
+        ) = decodeOrder(_data);
+
+        require(fromToken == ETH_ADDRESS, "order is not from ETH");
+
+        bytes32 key = _keyOf(
+            IERC20(ETH_ADDRESS),
```
Comment (nachomazzara) [inline contracts/UniswapEX.sol]: I would use `fromToken` for readability here.

File: contracts/UniswapEX.sol [hunk 2]
```diff
@@ -100,26 +121,43 @@ contract UniswapEX {
...
             _owner,
-            _salt,
+            _witness,
             amount
         );
     }
 
+    function readWitnesses(
+        bytes calldata _witnesses
+    ) external view returns (address) {
```
Comment (nachomazzara) [inline contracts/UniswapEX.sol]: Why `external`? Can we just set it as `internal`?

File: contracts/UniswapEX.sol [hunk 3]
```diff
@@ -100,26 +121,43 @@ contract UniswapEX {
...
         uint256 _minReturn,
         uint256 _fee,
         address payable _owner,
-        bytes32 _salt
+        bytes calldata _witnesses
     ) external {
+        // Calculate witness using signature
+        // avoid front-run by requiring msg.sender to know
+        // the secret
+        address witness = SigUtils.ecrecover2(
```
Comment (nachomazzara) [inline contracts/UniswapEX.sol]: Can we use `readWitnesses`? If not, I think that I would remove `readWitnesses`

**Comment by Agusx1211** [comment]
Mainnet contract: 0xbd2a43799b83d9d0ff56b85d4c140bce3d1d1c6c

**Pull request merged**